crates/common/tedge_utils/src/notify.rs

```diff
@@ -267,21 +329,109 @@ mod tests {
+            assert_rx_stream(expected_events, &mut stream).await;
         });
 
         let file_handler = tokio::task::spawn(async move {
             ttd_a_clone.file("file_a"); // should match CREATE
             ttd_b_clone.file("file_b").with_raw_content("content"); // should match MODIFY;
-            ttd_c_clone.file("file_c").delete(); // should match CREATE, DELETE file;
+            ttd_c_clone.file("file_c").delete_with_delay(); // should match CREATE, DELETE file;
             ttd_d_clone.dir("dir_d"); // should match CREATE directory;
         });
 
         fs_notify_handler.await.unwrap();
         file_handler.await.unwrap();
     }
+
+    /// Make sure that `FsEvent::Modify` is emitted for various operations.
+    ///
+    /// In order to not duplicate various operations, consumers may want to only listen to `Modify` events, which should
+    /// signify that the content of the file changed so it should be read again. However different editors and kinds of
+    /// operation may result in other events, like `Rename`, `Delete`, etc. We want these operations to emit `Modify` as
+    /// well, so that the consumers can only subscribe to single type of event and properly respond every time a file
+    /// they're watching changes.
+    #[tokio::test]
+    async fn modify_emitted_for_move_copy_create_delete() {
+        // Arrange
+        let dir = TempTedgeDir::new();
+        let nested_dir = dir.dir("nested");
+
+        let mut stream = fs_notify_stream(&[&nested_dir.path()]).unwrap();
+
+        // Test create
+        let file = nested_dir.file("file");
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
+            )]
+            .into(),
+            &mut stream,
+        )
+        .await;
+
+        // Test move to parent directory
+        std::fs::rename(
+            file.path(),
+            dir.path().join(file.path().file_name().unwrap()),
+        )
+        .unwrap();
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
+            )]
+            .into(),
+            &mut stream,
+        )
+        .await;
+
+        // Test copy to directory
+        std::fs::copy(
+            dir.path().join(file.path().file_name().unwrap()),
+            file.path(),
+        )
+        .unwrap();
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
+            )]
+            .into(),
+            &mut stream,
+        )
+        .await;
+
+        // Test remove
+        std::fs::remove_file(file.path()).unwrap();
+        assert_rx_stream(
+            [(
+                file.utf8_path().file_name().unwrap().to_string(),
+                vec![FsEvent::Modified],
```

Comment:
Here also, I was expecting the `FsEvent::FileDeleted` event right before the `FsEvent::Modified` event, as per the logic: `allow consumers to only monitor for `Modified` event if they want` implemented in that second `for` loop.

Reply:
https://github.com/thin-edge/thin-edge.io/pull/2454#discussion_r1397172580